src/libopenrave/kinbodygeometry.cpp

```diff
@@ -165,26 +162,52 @@ void AppendBoxTriangulation(const Vector& pos, const Vector& ex, TriMesh& tri)
     tri.indices.insert(tri.indices.end(), &indices[0], &indices[nindices]);
 }
 
-void AppendCylinderTriangulation(const Vector& pos, const dReal& rad, const dReal& len, const int& numverts, TriMesh& tri) {
+static void AppendConicalFrustumTriangulation(const Vector& pos, const dReal topRad, const dReal botRad, const dReal halfHeight, const uint numFaces, TriMesh& tri)
+{
     // once again, cylinder is on z axis
-    dReal dtheta = 2 * PI / (dReal)numverts;
-    int base = tri.vertices.size();
-    tri.vertices.push_back(Vector(0,0,len) + pos);
-    tri.vertices.push_back(Vector(0,0,-len) + pos);
-    tri.vertices.push_back(Vector(rad,0,len) + pos);
-    tri.vertices.push_back(Vector(rad,0,-len) + pos);
-    for(int i = 0; i < numverts+1; ++i) {
-        dReal s = rad * RaveSin(dtheta * (dReal)i);
-        dReal c = rad * RaveCos(dtheta * (dReal)i);
-        int off = (int)tri.vertices.size();
-        tri.vertices.push_back(Vector(c, s, len) + pos);
-        tri.vertices.push_back(Vector(c, s, -len) + pos);
-
-        tri.indices.push_back(base);       tri.indices.push_back(off-2);       tri.indices.push_back(off);
-        tri.indices.push_back(base+1);       tri.indices.push_back(off+1);       tri.indices.push_back(off-1);
-        tri.indices.push_back(off-2);   tri.indices.push_back(off-1);       tri.indices.push_back(off);
-        tri.indices.push_back(off);     tri.indices.push_back(off-1);       tri.indices.push_back(off+1);
-    }
+    const dReal dTheta = 2 * PI / (dReal)numFaces; // degrees to rotate every time
+
+    const int32_t base = tri.vertices.size();
+    tri.vertices.emplace_back(pos.x, pos.y, pos.z + halfHeight); // top center
+    tri.vertices.emplace_back(pos.x, pos.y, pos.z - halfHeight); // bottom center
+
+    for (uint i = 0; i < numFaces; ++i) {
+        const dReal unitX = RaveCos(dTheta * i);
+        const dReal unitY = RaveSin(dTheta * i);
+        const int32_t off = tri.vertices.size();
+
+        // line on the side
+        tri.vertices.emplace_back(pos.x + unitX * topRad, pos.y + unitY * topRad, pos.z + halfHeight); // line top
+        tri.vertices.emplace_back(pos.x + unitX * botRad, pos.y + unitY * botRad, pos.z - halfHeight); // line bottom
+
+        if (i > 0) {
+            // four triangles
+            tri.indices.insert(tri.indices.end(), {
```

Comment:
in interests of not implicitly allocating vectors, just so that we can use `insert`. please restore the old `push_back` methods. you can put your comments there. thanks

Reply:
in fact, unless, there's a bug, please don't make big changes to the implementation. it only increases review time. thanks

Reply:
this is not allocating vectors, it’s using initializer_list which is compile-time behavior.

Reply:
There is indeed an off-by-one bug in the original implementation. The first side edge is created twice (again at the end) causing this mesh not closing.

I did not change the implementation but merely refactored it (and added documentation).

Reply:
this is about defining the triangle mesh, and there are no repeating triangles. also i don't see any  "mesh not closing", you'll have to show it to me.
let's use `push_back`, you don't have to worry about saving code space.
thanks,

Reply:
Actually using indices is fastest (not push_back). Just timed both version,  your version is 2x+ slower. 
```
using indices: 6644[us]
using std::vector::insert: 14917[us]
```

You can see the perf script here:

https://pastebin.com/J5a8yq14


Can use this code:

```c++
void AppendCylinderTriangulationIndex(const Vector& pos, const dReal topRadius, const dReal bottomRadius, const dReal len, const int numverts, TriMesh& tri)
{
    // once again, cylinder is on z axis
    dReal dtheta = 2 * PI / (dReal)numverts;
    int base = tri.vertices.size();
    tri.vertices.resize(2*(numverts+1) + 4);
    tri.indices.resize((numverts+1) * 12);

    int vindex = 0;
    int iindex = 0;
    tri.vertices[vindex++] = Vector(0,0,len) + pos;
    tri.vertices[vindex++] = Vector(0,0,-len) + pos;
    tri.vertices[vindex++] = Vector(topRadius,0,len) + pos;
    tri.vertices[vindex++] = Vector(bottomRadius,0,-len) + pos;
    for(int i = 0; i < numverts+1; ++i) {
        dReal s = sinf(dtheta * (dReal)i);
        dReal c = cosf(dtheta * (dReal)i);
        int off = (int)tri.vertices.size();
        tri.vertices[vindex++] = Vector(topRadius*c, topRadius*s, len) + pos;
        tri.vertices[vindex++] = Vector(bottomRadius*c, bottomRadius*s, -len) + pos;

        tri.indices[iindex++] = (base);       tri.indices[iindex++] = (off-2);       tri.indices[iindex++] = (off);
        tri.indices[iindex++] = (base+1);       tri.indices[iindex++] = (off+1);       tri.indices[iindex++] = (off-1);
        tri.indices[iindex++] = (off-2);   tri.indices[iindex++] = (off-1);       tri.indices[iindex++] = (off);
        tri.indices[iindex++] = (off);     tri.indices[iindex++] = (off-1);       tri.indices[iindex++] = (off+1);
    }
}
```

Reply:
Reason it's faster because the `[]` operator skips bound checks. If you change it to `.at` it goes back to the same level as `.insert` or `.push_back`. Although no worry of the segfault here if the size is calculated correctly.

Also the code is incorrect as `.size()` now reports the full size, have to do more math there to calculate the vertices index, making the code harder to read and maintain.

Your decision though.

Reply:
Refactored into a even faster version.